core/state/caching_versa_db.go

```diff
@@ -0,0 +1,297 @@
+package state
+
+import (
+	"errors"
+	"fmt"
+	"sync/atomic"
+
+	versa "github.com/bnb-chain/versioned-state-database"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/lru"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/trie"
+	"github.com/ethereum/go-ethereum/trie/trienode"
+	"github.com/ethereum/go-ethereum/triedb"
+)
+
+type cachingVersaDB struct {
+	triedb        *triedb.Database
+	versionDB     versa.Database
+	codeDB        ethdb.KeyValueStore
+	codeSizeCache *lru.Cache[common.Hash, int]
+	codeCache     *lru.SizeConstrainedCache[common.Hash, []byte]
+
+	accTree  *VersaTree
+	state    versa.StateHandler
+	root     common.Hash
+	mode     versa.StateMode
+	hasState atomic.Bool
+}
+
+// NewVersaDatabase should be call by NewDatabaseWithNodeDB
+// TODO:: NewDatabaseWithNodeDB should add mode param.
+func NewVersaDatabase(db ethdb.Database, triedb *triedb.Database, mode versa.StateMode) Database {
+	return &cachingVersaDB{
+		triedb:        triedb,
+		versionDB:     triedb.VersaDB(),
+		codeDB:        db,
+		codeSizeCache: lru.NewCache[common.Hash, int](codeSizeCacheSize),
+		codeCache:     lru.NewSizeConstrainedCache[common.Hash, []byte](codeCacheSize),
+	}
+}
+
+func (cv *cachingVersaDB) OpenTrie(root common.Hash) (Trie, error) {
+	if cv.hasState.Load() {
+		//TODO:: will change to log.Error after stabilization
+		panic("account tree has open")
+	}
+
+	// TODO:: if root tree, versa db shouldb ignore check version
+	state, err := cv.versionDB.OpenState(0, root, cv.mode)
```

Comment:
Suggest to use the block num as the version here, not 0.

Reply:
It’s fine, but currently, this version is of no use to the upper layer and has brought about a burden. The plan is to make the upper layer aware of the version only if needed in the future. The conclusion is that it can be transparent to the upper layer for the time being.